glide-core/redis-rs/redis/src/cluster_routing.rs

```diff
@@ -243,14 +244,106 @@ pub fn combine_array_results(values: Vec<Value>) -> RedisResult<Value> {
-/// for each array of results, `sorting_order` should contain a matching array with the indices of
-/// the results in the final array.
-pub(crate) fn combine_and_sort_array_results<'a>(
+// An iterator that yields `Cow<[usize]>` representing grouped result indices according to a specified argument pattern.
+// This type is used to combine multi-slot array responses.
+type MultiSlotResIdxIter<'a> = std::iter::Map<
+    std::slice::Iter<'a, (Route, Vec<usize>)>,
+    fn(&'a (Route, Vec<usize>)) -> Cow<'a, [usize]>,
+>;
+
+/// Generates an iterator that returns grouped result indices for each slot in a multi-slot command response,
+/// structured according to the given argument pattern.
+///
+/// # Arguments
+/// * `route_arg_indices` - A reference to a vector where each element is a tuple containing a route and
+///   the corresponding argument indices for that route.
+/// * `args_pattern` - Specifies the argument pattern (e.g., `KeysOnly`, `KeyValuePairs`, ..), which defines how the indices are grouped for each slot.
+///
+/// # Returns
+/// An iterator yielding `Cow<[usize]>` with the grouped result indices based on the specified argument pattern.
+///
+/// /// For example, given the command `MSET foo bar foo2 bar2 {foo}foo3 bar3` with the `KeyValuePairs` pattern:
+/// - `route_arg_indices` would include:
+///   - Slot of "foo" with argument indices `[0, 1, 4, 5]` (where `{foo}foo3` hashes to the same slot as "foo" due to curly braces).
+///   - Slot of "foo2" with argument indices `[2, 3]`.
+/// - Using the `KeyValuePairs` pattern, each key-value pair contributes a single response, yielding three responses total.
+/// - Therefore, the iterator generated by this function would yield grouped result indices as follows:
+///   - Slot "foo" is mapped to `[0, 2]` in the final result order.
+///   - Slot "foo2" is mapped to `[1]`.
+fn multi_slot_sorting_order<'a>(
+    route_arg_indices: &'a [(Route, Vec<usize>)],
+    args_pattern: &MultiSlotArgPattern,
+) -> MultiSlotResIdxIter<'a> {
+    match args_pattern {
+        MultiSlotArgPattern::KeysOnly => route_arg_indices
+            .iter()
+            .map(|(_, indices)| Cow::Borrowed(indices)),
+        MultiSlotArgPattern::KeysAndPath => {
+            // The last index corresponds to the path, skip it
+            route_arg_indices
+                .iter()
+                .map(|(_, indices)| Cow::Borrowed(&indices[..indices.len() - 1]))
+        }
+        MultiSlotArgPattern::KeyPathValueTriples => {
+            // For each triplet (key, path, value) we receive a single response.
+            // For example, for argument indices: [(_, [0,1,2]), (_, [3,4,5,9,10,11]), (_, [6,7,8])]
+            // The resulting grouped indices would be: [0], [1, 3], [2]
+            route_arg_indices.iter().map(|(_, indices)| {
+                Cow::Owned(
+                    indices
+                        .iter()
+                        .step_by(3)
```

Comment:
we should probably add a check and see that array size divides by `3`

Reply:
done